feat: update azure function module and add tests

helayoty opened the pull request:
## All Submissions:
-------------------------------------
* [YES] Have you followed the guidelines in our Contributing [document](./CONTRIBUTING.md)?
* [YES] Have you added an explanation of what your changes do and why you'd like us to include them?
* [YES] I have updated the documentation accordingly.
* [YES] I have added tests to cover my changes.
* [YES] All new and existing tests passed.
* [YES] My code follows the code style of this project.
* [YES] I ran lint checks locally prior to submission.
* [YES] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

## What is the current behavior?
-------------------------------------
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

The current azure function module 
- wasn't stable when using with virtual network
- doesn't have unit/integration tests
- doesn't support managed identity

Issue Number: N/A


## What is the new behavior?
-------------------------------------
<!-- Please describe the behavior or changes that are being added by this PR. -->

- Use `azurerm_app_service_virtual_network_swift_connection` resource instead of `using azurerm_template_deployment`
- Simplify locals block. (it was so hard to use the old one)
- Update README 
- Add unit/integration tests 
- Simplify outputs and add managed identity output 

## Does this introduce a breaking change?
-------------------------------------
- [YES]

<!-- If this introduces a breaking change, please describe the impact and migration path for existing applications below. -->

All templates use the old module needs to update the usage and rerun the template tests
## Any relevant logs, error output, etc?
-------------------------------------
(If it’s long, please paste to https://ghostbin.com/ and insert the link here.)


## Other information
-------------------------------------
<!-- Any other information that is important to this PR such as screenshots of how the component looks before and after the change. -->

helayoty commented:
Cc: @nmiodice

ThePhen commented:
@helayoty can you elaborate on the utility of breaking the existing interface (renaming inputs and outputs). 

If the names were the same as before, would your (very desirable) improvements still make the module incompatible with existing templates?

helayoty commented:
> @helayoty great work, here.
> 
> Can you elaborate on the utility of breaking the existing interface (renaming inputs and outputs)?
> 
> If the names were the same as before, would your (very desirable) improvements still make the module incompatible with existing templates?

Hey @ThePhen , thanks for reviewing the PR. I agree with you some of the variables' names might be the same but this was the customer feedback that names of many variables/output are not meaningful so I agreed to change they especially I already changed the module core as well.  I'd be happy to apply this fix on the main template after the merge if this ok with you.

TechnicallyWilliams commented on infra/modules/providers/azure/function-app/README.md:
```diff
@@ -255,14 +72,21 @@ You will need to provide the following information for the dependencies:
 
 ### Manually deploying Private ACR Images
 
 If no image is deployed due to the usage of a Private ACR configuration then the `image` should be left blank and images can be deployed post terraform provisioning has occured by using the Azure CLI or some other process.
```
In addition to the important explanation here, we can be be a bit more clear and say something like, `"..or some other process. The below example uses the Azure CLI and docker."`

TechnicallyWilliams commented on infra/modules/providers/azure/function-app/main.tf:
```diff
@@ -1,89 +1,89 @@
-    // The value of "DOCKER" is a stand-in value that can be used to force the webapp created to be
-    // container compatible without explicitly specifying the image that the app should run.
-    config.image == "" ? "DOCKER" : format("DOCKER|%s/%s", var.docker_registry_server_url, config.image)
+    for config in values(var.fn_app_config) :
```
Suggestion: `for config in local.fn_configs `. I'm suggesting because line 34 proves that local variables can references other local variables.

TechnicallyWilliams commented on infra/modules/providers/azure/function-app/main.tf:
```diff
@@ -1,89 +1,89 @@
+  ]
+
+  app_deployment_config = [
+    for config in values(var.fn_app_config) :
```
Suggestion: Replace with `for config in local.fn_configs `. I'm suggesting because line 34 proves that local variables can references other local variables.

TechnicallyWilliams commented on infra/modules/providers/azure/function-app/main.tf:
```diff
@@ -1,89 +1,89 @@
   }
 
   lifecycle {
-    # This stanza will prevent terraform from reverting changes to the application container settings.
```
Is this comment about the stanza outdated? Why remove it?

TechnicallyWilliams commented on infra/modules/providers/azure/function-app/main.tf:
```diff
@@ -1,89 +1,89 @@
-    // container compatible without explicitly specifying the image that the app should run.
-    config.image == "" ? "DOCKER" : format("DOCKER|%s/%s", var.docker_registry_server_url, config.image)
+    for config in values(var.fn_app_config) :
+    config.image == null ? "" : format("DOCKER|%s/%s", var.docker_registry_server_url, config.image)
```
The hardcoded "DOCKER" string is no longer here. Has the logic for allowing "container" compatibility w/o an image been relocated?

TechnicallyWilliams commented on infra/modules/providers/azure/function-app/tests/unit/funtion_app_unit_test.go:
```diff
@@ -0,0 +1,81 @@
+package unit
```
Filename typo: `function_app_unit_test.go`

helayoty commented on infra/modules/providers/azure/function-app/terraform.tfvars.template:
```diff
@@ -0,0 +1,16 @@
+resource_group_name = ""
+resource_tags       = { "environment" = "dev" }
+service_plan_name   = ""
```
We can't retrieve the service plan by id according to [Data Source: azurerm_app_service_plan](https://www.terraform.io/docs/providers/azurerm/d/app_service_plan.html)

TechnicallyWilliams replied:
See comment above with suggestion to remove data source stanza.

helayoty commented on infra/modules/providers/azure/function-app/tests/test.tfvars.template:
```diff
@@ -0,0 +1,18 @@
+resource_group_name = ""
+resource_tags       = { "environment" = "dev" }
+service_plan_name   = "terraform-test-service-plan"
```
We can't retrieve the service plan by id according to [Data Source: azurerm_app_service_plan](https://www.terraform.io/docs/providers/azurerm/d/app_service_plan.html)

TechnicallyWilliams replied:
See comment above with suggestion to remove data source stanza.

helayoty commented on infra/modules/providers/azure/function-app/main.tf:
```diff
@@ -1,89 +1,89 @@
+  name = var.storage_account_resource_group_name
+}
+
+data "azurerm_app_service_plan" "main" {
```
Is it the ask here to let the user provides both name and id?

TechnicallyWilliams replied:
Did you consider discarding this data stanza completely and placing the burden of providing the `service_plan_id` on the caller of this module? Looks like the original approach was less complex. Consider keeping it.

TechnicallyWilliams replied:
Did you consider discarding this data stanza completely and instead allow the caller of this module to provide the `service_plan_id`? Looks like the original approach was less complex. Consider keeping it.

helayoty commented:
Thank you so much @TechnicallyWilliams  for your review. I've addressed all comments except 2 and waiting for your replying back. After that, if you can allow me to rebase.
